crates/oct-cloud/src/state.rs

```diff
@@ -13,17 +13,144 @@ pub struct Ec2InstanceState {
     pub instance_profile: Option<InstanceProfileState>,
 }
 
+impl Ec2InstanceState {
+    pub async fn new(ec2_instance: aws::Ec2Instance) -> Self {
+        Self {
+            id: ec2_instance.id.clone().expect("Instance id is not set"),
+            public_ip: ec2_instance
+                .public_ip
+                .clone()
+                .expect("Public ip is not set"),
+            public_dns: ec2_instance
+                .public_dns
+                .clone()
+                .expect("Public dns is not set"),
+            region: ec2_instance.region.clone(),
+            ami: ec2_instance.ami.clone(),
+            instance_type: ec2_instance.instance_type.clone().to_string(),
+            name: ec2_instance.name.clone(),
+            instance_profile: ec2_instance
+                .instance_profile
+                .as_ref()
+                .map(|profile| InstanceProfileState::new(profile)),
+        }
+    }
+
+    pub async fn build_from_state(&self) -> Result<aws::Ec2Instance, Box<dyn std::error::Error>> {
+        // Load AWS configuration
+        let region_provider = aws_sdk_ec2::config::Region::new(self.region.clone());
+        let config = aws_config::defaults(aws_config::BehaviorVersion::latest())
+            .region(region_provider)
+            .load()
+            .await;
+
+        let ec2_client = aws_sdk_ec2::Client::new(&config);
+        // Initialize instance profile
+        let instance_type: aws_sdk_ec2::types::InstanceType =
+            aws_sdk_ec2::types::InstanceType::from(self.instance_type.as_str());
+
+        let instance_profile = if let Some(profile_state) = &self.instance_profile {
+            Some(InstanceProfileState::build_from_state(profile_state).await?)
+        } else {
+            None
+        };
+
+        Ok(aws::Ec2Instance {
+            client: aws::Ec2::new(ec2_client),
+            id: Some(self.id.clone()),
+            public_ip: Some(self.public_ip.clone()),
+            public_dns: Some(self.public_dns.clone()),
+            region: self.region.clone(),
+            ami: self.ami.clone(),
+            instance_type,
+            name: self.name.clone(),
+            user_data: "".to_string(),
+            user_data_base64: "".to_string(),
+            instance_profile,
+        })
+    }
```

Comment:
You don't need this logic, it's already implemented in `aws::Ec2Instance::new`.

The same idea should be followed in the `InstanceProfileState` and `InstanceRoleState`